--- gstshmsink.c	2021-08-11 17:37:39.278034450 +0200
+++ tmp	2021-08-11 17:37:36.200360162 +0200
@@ -728,6 +728,13 @@
     while ((memory =
             gst_shm_sink_allocator_alloc_locked (self->allocator,
                 gst_buffer_get_size (buf), &self->params)) == NULL) {
+      GST_WARNING_OBJECT(self, "No free shared memory blocks, skip input buffer");
+      GST_OBJECT_UNLOCK(self);
+      return GST_FLOW_OK;
+
+      // TODO currently it can't allocate more memory then initial size, so next
+      // code don't work properly: instead of allocation new memory we have
+      // hanging in a while loop
       g_cond_wait (&self->cond, GST_OBJECT_GET_LOCK (self));
       if (self->unlock) {
         GST_OBJECT_UNLOCK (self);
